sycl/doc/extensions/proposed/sycl_ext_oneapi_user_defined_reductions.asciidoc

```diff
@@ -0,0 +1,220 @@
+and `joint_reduce` free functions defined in section 4.17.4.5. `reduce` of the
+core SYCL specification by allowing user-defined binary operators and
+non-fundamental types.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 Feature test macros. Therefore, an implementation
+supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_USER_DEFINED_REDUCTIONS` to one of the values defined in the
+table below.
+Application can test for existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's value
+to determine which of the extensions's APIs the implementation supports.
+
+Table 1. Values of the `SYCL_EXT_ONEAPI_USER_DEFINED_REDUCTIONS` macro.
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version. Base features are supported.
+|===
+
+=== Reduction functions
+
+This extension provides two overloads of `reduce_over_group` defined by the core
+SYCL specification.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+  template <typename GroupHelper, typename Ptr, typename BinaryOperation>
+  std::iterator_traits<Ptr>::value_type joint_reduce(GroupHelper g, Ptr first, Ptr last, BinaryOperation binary_op); // (1)
+
+  template <typename GroupHelper, typename Ptr, typename T, typename BinaryOperation>
+  T joint_reduce(GroupHelper g, Ptr first, Ptr last, T init, BinaryOperation binary_op); // (2)
+
+  template <typename GroupHelper, typename T, typename BinaryOperation>
+  T reduce_over_group(GroupHelper g, T x, BinaryOperation binary_op); // (3)
+
+  template <typename GroupHelper, typename V, typename T, typename BinaryOperation>
+  T reduce_over_group(GroupHelper g, V x, T init, BinaryOperation binary_op); // (4)
+}
+----
+
+**TODO**: consider adding `is_group_with_scratchpad_v` helper to describe
+preconditions better.
+
+1._Constraints_: `GroupHelper` is a `group_with_scratchpad`.
+
+_Mandates_: `binary_op(*first, *first)` must return a value of type
+`std::iterator_traits<Ptr>::value_type`.
+
+_Preconditions_: `first`, `last` and the type of `binary_op` must be the same
+for all work-items in the group. `binary_op` must be an instance of a function
+object.
+
+_Returns_: The result of combining the values resulting from dereferencing all
+iterators in the range `[first, last)` using the operator `binary_op`, where the
+values are combined according to the generalized sum defined in standard C++.
+
+NOTE: If `T` is a fundamental type and `BinaryOperation` is a SYCL function
+object type, then memory attached to `GroupHelper` object `g` is not used and
+the call to this overload is equivalent to calling
+`sycl::joint_reduce(g.get_group(), first, last, binary_op)`.
+
+2._Constraints_: `GroupHelper` is a `group_with_scratchpad`.
+
+_Mandates_: `binary_op(init, *first)` must return a value of type `T`.
```

Comment:
C++ allows this to return any type that is convertible to `T`.  Should we say the same?

Reply:
I think we could say the same, yes. I was simply copying those requirements from existing ones from [4.17.4.5. reduce](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_reduce). So, if we go this way, I suggest that we also update the core spec

Reply:
> I think we could say the same, yes. I was simply copying those requirements from existing ones from [4.17.4.5. reduce](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_reduce). So, if we go this way, I suggest that we also update the core spec

@gmlueck , what do you think? We aligned the interface of the functions with the core spec.

Reply:
@Pennycook has been driving this feature in the core spec.  I defer to his judgement on whether it should be changed.

Reply:
I'm not sure I remember why that restriction is there in the core spec.  I have vague recollections of talking to @rolandschulz about this -- if I recall correctly, we consciously went with "return a value of `T`" because we wanted a less surprising behavior than `std::reduce`.

@rolandschulz, does that sound right to you?  If we extend the interface here to support user-defined functors, I suspect libraries may want to implement `std::reduce` using `sycl::reduce_over_group`, in which case having different behavior may be undesirable.  Should we match the "convertible to `T`" behavior for user-defined functors, at least?

Reply:
@rolandschulz, friendly ping

Reply:
Yes the concern was that it would be easy to accidentally write: 
```
float f = ...;
float s = reduce_over_group(g, f, 0, plus<>);
```
And get unintentally an integer sum without diagnostic.

This concern also applies to user-defined function objects. And it would be weird to have the constraint be different for user-defined function objects.

Reply:
Thanks.  I think we should leave things as they are for now, then.  We can revisit if this decision if any users complain that the behavior is different.  So far nobody has mentioned this.